src/proposal.mligo

```diff
@@ -330,31 +327,93 @@ let handle_proposal_is_over
+                then (None : proposal_doubly_linked_list option)
+                else store.ongoing_proposals_dlist
+              )  in
+      ((ops, config, {store with ongoing_proposals_dlist = plist}, counter), should_cont)
+
+(*
+ * Meant to be as `fold_with_flush_one_rec (flush_one, key, plist, result)`
+ *
+ * Helper function for `fold_with_flush_one`
+ *)
+let rec fold_with_flush_one_rec
+  ( key, plist, result
+  : proposal_key * proposal_doubly_linked_list * plist_fold_acc
+  ) : plist_fold_acc =
+  match Map.find_opt (key, true) plist.map with
+    | Some next ->
+        // Apply `flush_one` on the key, continue looping until no key is found.
+        let (result, should_cont) = flush_one (result, (key, false)) in
+        if (should_cont = true) then fold_with_flush_one_rec (next, plist, result)
+        else result
+    | None ->
+        // In case of no key is found, check the last element and apply `flush_one` to it.
+        if plist.last = key
+          then let (result, _) = flush_one (result, (plist.last, true)) in
+            result
+          else (failwith bad_state : plist_fold_acc) // Shouldn't happen, next exist but no last.
+
+
+
+(*
+ * Meant to be as `fold (flush_one, plist, init)`.
+ * Cannot pass `flush_one` as f, since this result in complex compile errors.
+ *
+ * Apply `flush_one` on every proposal_key in `proposal_doubly_linked_list`. Exit early if the function return `false` in its result.
+ *)
+let fold_with_flush_one (plist_o, init: (proposal_doubly_linked_list option) * plist_fold_acc): plist_fold_acc =
+  match plist_o with
+    | None ->
+        // Empty plist, return the initial
+        init
+    | Some plist ->
+        // Apply `flush_one` on the first key
+        if (plist.first = plist.last) then
+          // Only 1 key exist.
+          let (result, _) = flush_one (init, (plist.first, true)) in
+          result
+        else
+          fold_with_flush_one_rec (plist.first, plist, init)
+
+
+(*
+ * Flush all proposals that passed their voting stage.
+ *)
 let flush(n, config, store : nat * config * storage): return =
   if n = 0n
   then (failwith empty_flush : return)
   else
     let counter : counter = { current = 0n; total = n } in
-    let flush_one
-        (acc, e: (operation list * storage * counter) * (blocks * proposal_key)) =
-          let (ops, store, counter) = acc in
-          let (start_level, proposal_key) = e in
-          handle_proposal_is_over (config, start_level, proposal_key, store, ops, counter)
-        in
-    let (ops, store, counter) =
-      Set.fold flush_one store.proposal_key_list_sort_by_level (nil_op, store, counter)
+    let init = (nil_op, config, store, counter) in
+
+    let (ops, _, store, counter) =
+      fold_with_flush_one (store.ongoing_proposals_dlist, init)
     in
     // prevent empty flushes to avoid gas costs when unnecessary.
     if counter.current = 0n
     then (failwith empty_flush : return)
     else (ops, store)
```

Comment:
I see that this has gotten a lot more complex, couldn't it be handled with a single recursive function?

Such function would:
1. check if the number `n` was reached or the dlist is empty, return if so
2. `pop` the first element off of the dlist (updating the dlist in the storage)
3. `handle_proposal_is_over` on that first element (possibly failing)
4. repeat with the updated store and `n` decreased by one

Does that sound correct or am I missing something?

Reply:
This definitely simplifies thing a lot. The previous implementation try to keep `handle_proposal_is_over` the same, but I think your suggestion is better.

I've updated this now but I slightly modify this by adding a check to ensure that `plist.first` is able to be flushed in the stage 1.

Reply:
I think that this is much better, I have added some suggestion to tweak it a little bit more, let me know what do you think about those.

Reply:
@pasqu4le I've made all the changes according to your suggestions, also update the haskell SMT and tests. This should be ready for re-review.